sdk/core/Azure.Core/src/Pipeline/BearerTokenAuthenticationPolicy.cs

```diff
@@ -88,5 +113,76 @@ private async ValueTask ProcessAsync(HttpMessage message, ReadOnlyMemory<HttpPip
                 ProcessNext(message, pipeline);
             }
         }
+
+        private string? GetConcurrentlySetHeaderValue()
+        {
+            lock (_syncObj)
+            {
+                _headerSemaphore?.Release(1);
+                return _headerValue;
+            }
+        }
+
+        private string? SetHeaderValue(in AccessToken token)
+        {
+            var offset = _credential.TokenRefreshOffset;
+            lock (_syncObj)
+            {
+                _headerValue = "Bearer " + token.Token;
+                _refreshOn = token.ExpiresOn - offset;
+                _expiresOn = token.ExpiresOn;
+                _tokenState = TokenState.Valid;
+
+                _headerSemaphore?.Release(1);
+                return _headerValue;
+            }
+        }
+
+        private (string? headerValue, SemaphoreSlim? semaphore) UpdateTokenState()
+        {
+            lock (_syncObj)
+            {
+                switch (_tokenState) {
+                    case TokenState.Invalid:
+                        _tokenState = TokenState.Pending;
+                        ResetSemaphore(_headerSemaphore);
+                        return (null, null);
+                    case TokenState.Pending:
+                        _headerSemaphore ??= new SemaphoreSlim(0, 1);
```

Comment:
Using convenience operators list ??= here and the ternary operator below can obscure synchronization issues in code which is intended to be thread safe, as what appears to be one statement is actually not executed atomically. I realize that this code is all under a lock so it's not really an issue, but if the code might be refactored in the future perhaps it's worth being explicit?

Reply:
`_headerSemaphore` should never be accessed outside of the lock. But maybe it will go away cause I'm thinking about replacing it with a different primitive.

Reply:
Extracted this logic into separate type, `AccessTokenCache`.